src/paths/random_walk.c

```diff
@@ -32,86 +32,49 @@
- *   stuck, then the \p stuck argument specifies what happens. The
- *   starting vertex counts as one step.
- * \param mode How to walk along the edges in directed graphs.
- *   \c IGRAPH_OUT means following edge directions, \c IGRAPH_IN means
- *   going opposite the edge directions, \c IGRAPH_ALL means ignoring
- *   edge directions. This argument is ignored for undirected graphs.
- * \param stuck What to do if the random walk gets stuck.
- *   \c IGRAPH_RANDOM_WALK_STUCK_RETURN means that the function returns
- *   with a shorter walk; \c IGRAPH_RANDOM_WALK_STUCK_ERROR means
- *   that an error is reported. In both cases \p walk is truncated
- *   to contain the actual interrupted walk.
- * \return Error code.
- *
- * Time complexity: O(l + d), where \c l is the length of the
- * walk, and \c d is the total degree of the visited nodes.
+ * This function performs a random walk with a given length on a graph,
+ * from the given start vertex.
+ * It's used for igraph_random_walk when the given graph is unweighted,
+ * and only vertex IDs of the vertices on the walk are needed (edge IDs are not needed).
+ * \param vertices An allocated vector, the result is stored here as
+ *   a list of vertex IDs. It will be resized as needed.
+ *   It includes the starting vertex id as well.
  */
-
-
-igraph_error_t igraph_random_walk(const igraph_t *graph, igraph_vector_int_t *walk,
-                       igraph_integer_t start, igraph_neimode_t mode,
-                       igraph_integer_t steps,
-                       igraph_random_walk_stuck_t stuck) {
-
-    /* TODO:
-       - multiple walks potentially from multiple start vertices
-       - weights
-    */
-
-    igraph_lazy_adjlist_t adj;
-    igraph_integer_t vc = igraph_vcount(graph);
+static igraph_error_t igraph_i_random_walk_adjlist(const igraph_t *graph,
+                                    igraph_vector_int_t *vertices,
+                                    igraph_integer_t start,
+                                    igraph_neimode_t mode,
+                                    igraph_integer_t steps,
+                                    igraph_random_walk_stuck_t stuck) {
     igraph_integer_t i;
-
-    if (start < 0 || start >= vc) {
-        IGRAPH_ERRORF("Starting vertex must be between 0 and the "
-                      "number of vertices in the graph (%" IGRAPH_PRId
-                      "), got %" IGRAPH_PRId ".", IGRAPH_EINVAL,
-                      vc, start);
-    }
-    if (steps < 0) {
-        IGRAPH_ERRORF("Number of steps should be non-negative, got %"
-                      IGRAPH_PRId ".", IGRAPH_EINVAL, steps);
-    }
-
+    igraph_lazy_adjlist_t adj;
+    
     IGRAPH_CHECK(igraph_lazy_adjlist_init(graph, &adj, mode, IGRAPH_LOOPS, IGRAPH_MULTIPLE));
     IGRAPH_FINALLY(igraph_lazy_adjlist_destroy, &adj);
 
-    IGRAPH_CHECK(igraph_vector_int_resize(walk, steps));
+    IGRAPH_CHECK(igraph_vector_int_resize(vertices, steps));
```

Comment:
```suggestion
    IGRAPH_CHECK(igraph_vector_int_resize(vertices, steps + 1));
```

Right?

Reply:
Yes, you're right! Also, I need to change the for loop on l. 60 from
```C
for (i = 1; i < steps; i++)
```
to 
```C
for (i = 1; i <= steps; i++)
```

Reply:
Ah, yes, indeed, good catch!